ocaml/xapi/xapi_host.ml

```diff
@@ -596,48 +611,104 @@ let evacuate ~__context ~host ~network =
+  let assert_valid_networks plans =
+    plans
+    |> Hashtbl.to_seq
+    |> Seq.filter_map (function _, Migrate host -> Some host | _ -> None)
+    |> List.of_seq
+    |> List.sort_uniq compare
+    |> List.iter (fun host ->
+           ignore
+           @@ Xapi_network_attach_helpers
+              .assert_valid_ip_configuration_on_network_for_host ~__context
+                ~self:network ~host
+       )
+  in
+
+  let options =
+    match network with
+    | network when network = Ref.null ->
+        [("live", "true")]
+    | network ->
+        assert_valid_networks plans ;
+        [("network", Ref.string_of network); ("live", "true")]
+  in
+
+  let migrate_vm ~rpc ~session_id (vm, plan) =
     match plan with
     | Migrate host ->
-        ( try
-            ( if network <> Ref.null then
-                let hosts = Db.Host.get_all ~__context in
-                List.iter
-                  (fun host ->
-                    ignore
-                    @@ Xapi_network_attach_helpers
-                       .assert_valid_ip_configuration_on_network_for_host
-                         ~__context ~self:network ~host
-                  )
-                  hosts
-            ) ;
-            let with_network_option =
-              if network <> Ref.null then
-                [("network", Ref.string_of network)]
-              else
-                []
-            in
-            let options = ("live", "true") :: with_network_option in
-            Helpers.call_api_functions ~__context (fun rpc session_id ->
-                Client.Client.VM.pool_migrate ~rpc ~session_id ~vm ~host
-                  ~options
-            )
-          with
-        | Api_errors.Server_error (code, _)
-          when code = Api_errors.vm_bad_power_state ->
-            ()
-        | e ->
-            raise e
-        ) ;
-        let progress = Db.Task.get_progress ~__context ~self:task in
-        TaskHelper.set_progress ~__context (progress +. individual_progress)
+        Client.Client.Async.VM.pool_migrate ~rpc ~session_id ~vm ~host ~options
     | Error (code, params) ->
         (* should never happen *)
         raise (Api_errors.Server_error (code, params))
   in
-  let () = Hashtbl.iter migrate_vm plans in
+
+  (* execute [n] asynchronous API calls [api_fn] for [xs] and wait for them to
+     finish before executing the next batch. *)
+  let batch ~__context n api_fn xs =
+    let finally = Xapi_stdext_pervasives.Pervasiveext.finally in
+    let destroy = Client.Client.Task.destroy in
+    let fail task msg =
+      raise
+        Api_errors.(
+          Server_error
+            (internal_error, [Printf.sprintf "%s, %s" (Ref.string_of task) msg])
+        )
+    in
+
+    let assert_success task =
+      match Db.Task.get_status ~__context ~self:task with
+      | `success ->
+          ()
+      | `failure -> (
+        match Db.Task.get_error_info ~__context ~self:task with
+        | [] | [_] ->
```

Comment:
Are we sure that `[_]` belongs here? couldn't it be an error message with 0 params (if such errors exist?), i.e. `code :: []`